Update to using mypy 1.15

aorenste opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #154054

The BC break isn't real - mypy decided to start complaining about the way we were typing that function.

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

aorenste committed:
mypy 1.15

[ghstack-poisoned]

Skylion007 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -1260,7 +1260,7 @@ def impure_pred(n: fx.Node) -> bool:
 
 def wrap_key(
-    f: Callable[_P, R], tensors: _P.args, tracer: _ProxyTracer, pre_dispatch: bool
+    f: Callable[_P, R], tensors: _P.args, tracer: _ProxyTracer, pre_dispatch: bool  # type: ignore[valid-type]
```
These are actual bugs using TypeVarTuple, either expand P.kwargs too or maybe use a TypeVarTuple if you only care about positional args, or type Tensors explicitly.

Skylion007 replied:
```suggestion
    f: Callable[Unpack[_Ts], R], 
    tensors: Tuple[Unpack[_Ts]],
    tracer: _ProxyTracer,
    pre_dispatch: bool,
```
I might have the syntax slightly off for the backwards compatible stuff with TypeVarTuple

Skylion007 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -2318,7 +2318,7 @@ def get_proxy_mode() -> Optional[ProxyTorchDispatchMode]:
 
 
-def handle_sym_dispatch(func: Callable[_P, R], args: _P.args, kwargs: _P.kwargs) -> R:
+def handle_sym_dispatch(func: Callable[_P, R], args: _P.args, kwargs: _P.kwargs) -> R:  # type: ignore[valid-type]
```
Okay yeah, you could as a TypeVarTuple for the positional args, but this doesn't allow the kwargs to be typed specifcially, if you can change the call signature to 
```suggestion
def handle_sym_dispatch(func: Callable[_P, R], *args: _P.args, **kwargs: _P.kwargs) -> R:
```
Typing will work properly
If you can't do that this will at least type check some of the positional args
```suggestion
_Ts = TypeVarTuple("_Ts")

def handle_sym_dispatch(func: Callable[_P, R], args: _Ts, kwargs: dict[str, Any]) -> R:
```

aorenste replied:
Using a TypeVar doesn't really add anything here - for that it might as well be `tuple[Any, ...]`.  I think I'd prefer leaving the type: ignore since at least then I can leave the _P.args/_P.kwargs dropping as documentation of what is intended to be there.

Skylion007 replied:
@aorenste it can preserve the tuple arguments to the next function call right? My understanding is the type var would preserve the function annotations

aorenste replied:
(Assuming I'm understanding what you're saying properly) I don't think so - I don't believe the type-checkers "dig through" functions like that.

When analyzing a function they only look at that function's own decl and the decls of called functions to determine whether the parameters are valid. So a lone 'T' doesn't give any information.  If you're calling `func(args)` then it only sees that `_P` and `T` are unrelated - it won't dig through how this function is being called to see if it fits.

Skylion007 reviewed (commented):
Use TypeVarTuple for wrap_key

aorenste committed:
Update on "mypy 1.15"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste commented:
@pytorchbot rebase

aorenste committed:
Update

[ghstack-poisoned]

aorenste committed:
Update on "Update to using mypy 1.15"


The BC break isn't real - mypy decided to start complaining about the way we were typing that function.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "Update to using mypy 1.15"


The BC break isn't real - mypy decided to start complaining about the way we were typing that function.

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste commented:
@pytorchbot merge
